### PR TITLE
Move database operations of merging a pull request to post receive hook and add a transaction

### PR DESCRIPTION
**Pull request by lunny**
Merging PR may fail because of various problems. The pull request may have a dirty state because there is no transaction when merging a pull request. ref https://github.com/go-gitea/gitea/pull/25741#issuecomment-2074126393

This PR moves all database update operations to post-receive handler for merging a pull request and having a database transaction. That means if database operations fail, then the git merging will fail, the git client will get a fail result.

There are already many tests for pull request merging, so we don't need to add a new one.

### CODE REVIEW

File: routers/private/hook_post_receive.go [hunk 1]
```diff
@@ -172,13 +184,16 @@ func HookPostReceive(ctx *gitea_context.PrivateContext) {
...
+		if pusher == nil {
+			var err error
+			pusher, err = user_model.GetUserByID(ctx, opts.UserID)
+			if err != nil {
+				log.Error("Failed to Update: %s/%s Error: %v", ownerName, repoName, err)
+				ctx.JSON(http.StatusInternalServerError, private.HookPostReceiveResult{
+					Err: fmt.Sprintf("Failed to Update: %s/%s Error: %v", ownerName, repoName, err),
+				})
+				return
+			}
```
Comment (yp05327) [inline routers/private/hook_post_receive.go]: It seems that this is not necessary. The logic is a bit strange here.

Comment (lunny) [reply routers/private/hook_post_receive.go]: 188e55b0e0e12710bd2b01774e2fbe987b658828

File: modules/repository/env.go [hunk 2]
```diff
@@ -25,11 +25,16 @@ const (
...
 	EnvDeployKeyID  = "GITEA_DEPLOY_KEY_ID"
 	EnvPRID         = "GITEA_PR_ID"
+	EnvPRAction     = "GITEA_PR_ACTION"
 	EnvIsInternal   = "GITEA_INTERNAL_PUSH"
 	EnvAppURL       = "GITEA_ROOT_URL"
 	EnvActionPerm   = "GITEA_ACTION_PERM"
 )
 
+const (
+	PullRequestActionMerge = "merge"
```
Comment (wxiaoguang) [inline modules/repository/env.go]: This design seems fragile and unclear.

File: services/pull/merge.go [hunk 3]
```diff
@@ -245,7 +236,7 @@ func Merge(ctx context.Context, pr *issues_model.PullRequest, doer *user_model.U
 }
 
 // doMergeAndPush performs the merge operation without changing any pull information in database and pushes it up to the base repository
-func doMergeAndPush(ctx context.Context, pr *issues_model.PullRequest, doer *user_model.User, mergeStyle repo_model.MergeStyle, expectedHeadCommitID, message string) (string, error) {
+func doMergeAndPush(ctx context.Context, pr *issues_model.PullRequest, doer *user_model.User, mergeStyle repo_model.MergeStyle, expectedHeadCommitID, message string, isMerge bool) (string, error) {
```
Comment (wxiaoguang) [inline services/pull/merge.go]: Why it is called `isMerge`? This function `doMergeXxx`, why isMerge could be false?

(I have read the code and I can answer the question, but, I do not think such naming and design is right)

**Comment by wxiaoguang** [comment]
Any test code?

File: services/pull/merge.go (hunk 3 above)
Comment (lunny) [reply services/pull/merge.go]: I think the original function name `doMergeAndPush` should be `doCommitAndPush`.

File: modules/repository/env.go [hunk 4]
```diff
@@ -25,11 +25,16 @@ const (
 	EnvKeyID        = "GITEA_KEY_ID" // public key ID
 	EnvDeployKeyID  = "GITEA_DEPLOY_KEY_ID"
 	EnvPRID         = "GITEA_PR_ID"
+	EnvPRAction     = "GITEA_PR_ACTION"
```
Comment (wxiaoguang) [inline modules/repository/env.go]: I think it should be called `GITEA_PUSH_TRIGGER`

* GITEA_PUSH_TRIGGER= (empty, default)
* GITEA_PUSH_TRIGGER=pr-update-with-base
* GITEA_PUSH_TRIGGER=pr-merge-to-base

etc

File: services/pull/merge.go (hunk 3 above)
Comment (wxiaoguang) [reply services/pull/merge.go]: But I can see this function just does "merge and push", see below: createTemporaryRepoForMerge, switch mergeStyle

File: modules/repository/env.go (hunk 4 above)
Comment (lunny) [reply modules/repository/env.go]: 3e320f865f1c93e1ce646da74f5cacc6c404e3e1

File: services/pull/merge.go (hunk 3 above)
Comment (lunny) [reply services/pull/merge.go]: 3e320f865f1c93e1ce646da74f5cacc6c404e3e1

**Comment by lunny** [comment]
> Any test code?

There are already enough tests for merging a pull request. And it's difficult to do a mock of updating database failure.

**Comment by wxiaoguang** [comment]
I am sure `handlePullRequestMerging` should and could be tested.

**Comment by wxiaoguang** [comment]
Write and add some tests

https://github.com/wxiaoguang/gitea/tree/lunny/transaction_merge

**Comment by lunny** [comment]
Please help to send to my branch and I have a comment in your branch that the `updates` can be great than 1 because of different merge styles.

**Comment by wxiaoguang** [comment]
> Please help to send to my branch and I have a comment in your branch that the `updates` can be great than 1 because of different merge styles.

No write permission. You could merge it.

**Comment by wxiaoguang** [comment]
> that the `updates` can be great than 1 because of different merge styles.

But you said so:

image

**Comment by lunny** [comment]
> > Please help to send to my branch and I have a comment in your branch that the `updates` can be great than 1 because of different merge styles.
> 
> No write permission. You could merge it.

It's wired I have allowed maintainers to edit the pull request.
image

**Comment by wxiaoguang** [comment]
> The comment is not right. My previous code is right. :)

I also do not think the comment is right. I just show you what's your comment's behavior. So you could correct the comment and code together.

**Comment by lunny** [comment]
@wxiaoguang Tests for auto-merge deletion added. 03ab2fa2711c22d55a4ddbfcc08fbeca701950eb

**Review by wxiaoguang — APPROVED** [review]

**Review by wolfogre — APPROVED** [review]

**Pull request merged**